### DISCUSSION TITLE
Authentication fails for some SSO users

### DISCUSSION THREAD

**Discussion by phidah** [outline/outline#4182] [Self Hosting]
I've set up Outline with Okta OpenID connect authentication, which works fine.
In Okta we have users with multiple email domains. For users with our primary domain the authentication works file. For users with another email domain, they get the following error in Outline when logging on (URL is `<domain>/?notice=invalid-authentication`):

> Authentication failed - you do not have permission to access this team.

The log shows the following (removed requests to static assets):

```
Sep 27 09:36:25 AM  {"label":"http","level":"info","message":"  <-- GET /auth/oidc"}
Sep 27 09:36:25 AM  {"label":"http","level":"info","message":"  --> GET /auth/oidc 302 3ms 0b"}
Sep 27 09:36:27 AM  {"error":"Invalid authentication","stack":"UnauthorizedError: Invalid authentication\n    at InvalidAuthenticationError (/opt/outline/build/server/errors.js:48:34)\n    at userProvisioner (/opt/outline/build/server/commands/userProvisioner.js:156:50)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:85:20)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:87:22)","level":"error","message":"Error during authentication"}
Sep 27 09:36:27 AM  {"label":"http","level":"info","message":"  --> GET /auth/oidc.callback?code=XXXXXXX&state=YYYYYYYYYYY 302 851ms 163b"}
Sep 27 09:36:27 AM  {"label":"http","level":"info","message":"  <-- GET /static/vendors~document~shared-document~shared-editor.1e8dc5831b40c6d105d3.js.LICENSE.txt"}
Sep 27 09:36:27 AM  {"label":"http","level":"info","message":"  --> GET /static/vendors~document~shared-document~shared-editor.1e8dc5831b40c6d105d3.js.LICENSE.txt 200 3ms 193b"}
Sep 27 09:36:27 AM  {"label":"http","level":"info","message":"  <-- GET /?notice=invalid-authentication"}
Sep 27 09:36:27 AM  {"label":"http","level":"info","message":"  --> GET /?notice=invalid-authentication 200 3ms 1.07kb"}
Sep 27 09:36:27 AM  {"label":"http","level":"info","message":"  <-- POST /api/auth.config"}
Sep 27 09:36:27 AM  {"label":"http","level":"info","message":"  --> POST /api/auth.config 200 7ms 129b"}
```

The log from Okta shows that the user has successfully been granted access:

```
OIDC access token is granted SUCCESS (User)
Access Token  (access_token)
```
This is the same log output I see when users successfully sign in.

I have tried adding all the possible domains under "Allowed domains" in Outline settings. I've also tried removing all domains (which I think would allow all domains - but it is unclear from the text how the feature works).

How would I best go about debugging this?

**Comment by Amahr** [comment]
I have the same problem with Authentik and Outline.
No domains added, so all domains should be allowed.

Inviting the the user by his email (used with Authentik) through Outline will enable the User to setup his account successfully with SSO.

**Reply by phidah** [reply]
Same here. Explicitly inviting the user solves the issue.

**Comment by tommoor** [comment]
Which version of Outline are you running?

**Reply by phidah** [reply]
We run 0.66.1

**Reply by Amahr** [reply]
Same here, version is v0.66.1

**Comment by Senorsen** [comment]
I think I may find the cause. In summary, all your user emails' domains from the same SSO provider should be the same (with the first logged-in user who is likely the system admin), such as xxx@example.com, otherwise, some users with different email domains (such as xxx@otherdomain.com) will face this authentication fail issue.

I mean no offense, but I think the cause of the strange behavior might be some SSO authentication code (user/account/team-Provisioner.js) may be implemented in the wrong way. :) Because authentication and SSO providers should be independent of the user emails.

Version 0.66.1

@tommoor

**Reply by almereyda** [reply]
In case you are using Keycloak as SSO provider, there is a known workaround on the side of Keycloak which normalises the email field to a common domain:

- https://github.com/outline/outline/issues/2678#issuecomment-987366852

**Comment by tommoor** [comment]
You can follow this issue: https://github.com/outline/outline/issues/4275
